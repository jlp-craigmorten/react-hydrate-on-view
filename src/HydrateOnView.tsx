import { Suspense, useEffect, useRef, useState } from 'react';

import type { FC, PropsWithChildren } from 'react';

interface SuspendConditionallyProps {
  suspend: boolean;
}

const SuspendConditionally: FC<
  PropsWithChildren<SuspendConditionallyProps>
> = ({ suspend, children }) => {
  if (suspend && typeof window !== 'undefined') {
    throw new Promise(() => {});
  }

  return <>{children}</>;
};

export interface HydrateOnViewProps {
  /**
   * Margin around the viewport. Can have values similar to the CSS margin property,
   * e.g. `10px 20px 30px 40px` (top, right, bottom, left).
   */
  rootMargin?: string;
  /**
   * Number between `0` and `1` indicating the percentage of the content that
   * should be visible before triggering hydration.
   */
  threshold?: number;
}

export const HydrateOnView: FC<PropsWithChildren<HydrateOnViewProps>> = ({
  children,
  rootMargin,
  threshold,
}) => {
  const ref = useRef<HTMLDivElement>(null);
  const [inView, setInView] = useState(false);

  useEffect(() => {
    if (inView) {
      return;
    }

    const observer = new IntersectionObserver(
      ([entry]) => {
        if (entry.isIntersecting) {
          observer.disconnect();
          setInView(true);
        }
      },
      {
        rootMargin,
        threshold,
      },
    );

    observer.observe(ref.current!);

    return () => {
      observer.disconnect();
    };
  }, [inView, rootMargin, threshold]);

  return (
    <div ref={ref}>
      <Suspense fallback={null}>
        <SuspendConditionally suspend={!inView}>
          {children}
        </SuspendConditionally>
      </Suspense>
    </div>
  );
};
