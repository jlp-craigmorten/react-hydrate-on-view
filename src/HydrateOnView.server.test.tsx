import { renderToString } from 'react-dom/server';
import { describe, expect, it } from 'vitest';

import { HydrateOnView } from './HydrateOnView';

describe('HydrateOnView Server-side', () => {
  it('should render all content when server-side rendered', () => {
    const html = renderToString(
      <HydrateOnView>
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    expect(html).toMatchSnapshot();
  });

  it('should render all content when server-side rendered with custom props', () => {
    const html = renderToString(
      <HydrateOnView
        rootMargin="10px 20px 30px 40px"
        threshold={0.5}
        style={{ background: 'red' }}
      >
        <div data-testid="child">Test Content</div>
      </HydrateOnView>,
    );

    expect(html).toMatchSnapshot();
  });
});
