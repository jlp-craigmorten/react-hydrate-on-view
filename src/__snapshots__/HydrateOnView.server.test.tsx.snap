// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HydrateOnView Server-side > should render all content when server-side rendered 1`] = `"<div><!--$--><div data-testid="child">Test Content</div><!--/$--></div>"`;

exports[`HydrateOnView Server-side > should render all content when server-side rendered with custom props 1`] = `"<div><!--$--><div data-testid="child">Test Content</div><!--/$--></div>"`;
